src/kvm/hypervisor.rs

```diff
@@ -431,28 +439,136 @@ impl IoRegionFd {
+            Ok(t) => t,
+        };
+
+        let mut ioregion = self.ioregion;
+        ioregion.rfd = -1;
+        ioregion.wfd = -1;
+
+        if let Err(e) = self.hv_mem.write(&ioregion) {
+            warn!(
+                "IoRegionFd: Could not write to HvMem while dropping IoRegionFd: {}",
+                e
+            );
+            return;
+        }
+
+        match tracee.vm_ioctl_with_ref(ioctls::KVM_SET_IOREGION(), &self.hv_mem) {
+            Err(e) => warn!("IoRegionFd: kvm ioregionfd ioctl injection failed: {}", e),
+            Ok(ret) => {
+                if ret != 0 {
+                    warn!("IoRegionFd: kvm ioregionfd remove syscall failed: {}", ret);
+                }
+            }
+        }
+
+        match tracee.close(self.hv_rf_hv) {
+            Err(e) => warn!("IoRegionFd: close injection failed: {}", e),
+            Ok(ret) => {
+                if ret != 0 {
+                    warn!(
+                        "IoRegionFd: failed to close hv_rf_hv in hypervisor: {}",
+                        ret
+                    )
+                }
+            }
+        }
+
+        match tracee.close(self.hv_wf_hv) {
+            Err(e) => warn!("IoRegionFd: close injection failed: {}", e),
+            Ok(ret) => {
+                if ret != 0 {
+                    warn!(
+                        "IoRegionFd: failed to close hv_wf_hv in hypervisor: {}",
+                        ret
+                    )
+                }
+            }
+        }
+
+        if let Err(e) = close(self.rf_hv) {
+            warn!("IoRegionFd: failed to close rf_hv: {}", e)
+        }
+
+        if let Err(e) = close(self.wf_hv) {
+            warn!("IoRegionFd: failed to close wf_hv: {}", e)
+        }
+
+        if let Err(e) = close(self.rfile) {
+            warn!("IoRegionFd: failed to close rfile: {}", e)
+        }
+
+        if let Err(e) = close(self.wfile) {
+            warn!("IoRegionFd: failed to close wfile: {}", e)
+        }
+    }
+}
+
+/// A handle implementing clone, read and write for the non-clonable IoRegionFd.
+#[derive(Debug, Clone)]
+pub struct RawIoRegionFd {
+    rfile: RawFd, // our end: we write responses here
+    wfile: RawFd, // we read commands from here
+    pollfds: Vec<PollFd>,
+    pub ioregion: kvm_ioregion,
+}
 
+impl RawIoRegionFd {
     /// receive read and write events/commands
-    pub fn read(&self) -> Result<ioregionfd_cmd> {
+    pub fn read(&mut self) -> Result<Option<ioregionfd_cmd>> {
         let len = size_of::<ioregionfd_cmd>();
         let mut t_mem = MaybeUninit::<ioregionfd_cmd>::uninit();
         // safe, because slice.len() == len
         let t_slice = unsafe { std::slice::from_raw_parts_mut(t_mem.as_mut_ptr() as *mut u8, len) };
+
+        // read
+        let timeout = TimeSpec::from(Duration::from_millis(100));
+        let nr_events = try_with!(
+            ppoll(&mut self.pollfds, Some(timeout), SigSet::empty()),
+            "read/ppoll failed"
+        );
```

Comment:
poll with timeout instead of read to allow for regular `should_stop` checks in ioregion-handler thread. Not sure if this is sufficiently good style.

Reply:
The only better way I see is using iouring to save one syscall...
You could probably increase the timeout to 300ms without scarifying user experience much.